datumaro/plugins/common_semantic_segmentation_format.py

```diff
@@ -0,0 +1,150 @@
+            label_map = parse_meta_file(path)
+            self._categories = make_categories(label_map)
+        else:
+            meta_file = glob.glob(osp.join(path, "**", DATASET_META_FILE), recursive=True)
+            if is_meta_file(meta_file[0]):
+                self._root_dir = osp.dirname(meta_file[0])
+
+                label_map = parse_meta_file(meta_file[0])
+                self._categories = make_categories(label_map)
+            else:
+                raise DatasetImportError("Dataset meta info file was not found in %s" % path)
+
+        self._items = list(self._load_items().values())
+
+    def _load_items(self):
+        items = {}
+
+        image_dir = osp.join(self._root_dir, CommonSemanticSegmentationPath.IMAGES_DIR)
+
+        if osp.isdir(image_dir):
+            images = {
+                osp.splitext(osp.relpath(p, image_dir))[0].replace("\\", "/")[
+                    len(self._image_prefix) :
+                ]: p
+                for p in find_images(image_dir, recursive=True)
+                if osp.basename(p).startswith(self._image_prefix)
+            }
+        else:
+            images = {}
+
+        mask_dir = osp.join(self._root_dir, CommonSemanticSegmentationPath.MASKS_DIR)
+        masks = [
+            mask_path
+            for mask_path in find_images(mask_dir, recursive=True)
+            if osp.basename(mask_path).startswith(self._mask_prefix)
+        ]
+
+        for mask_path in masks:
+            item_id = osp.splitext(osp.basename(mask_path))[0][len(self._mask_prefix) :]
+
+            image = images.get(item_id)
+            if image:
+                image = Image(path=image)
+
+            annotations = []
+            mask = lazy_mask(mask_path, self._categories[AnnotationType.mask].inverse_colormap)
+            mask = mask()  # loading mask through cache
+
+            classes = np.unique(mask)
+            for label_id in classes:
+                annotations.append(
+                    Mask(image=self._lazy_extract_mask(mask, label_id), label=label_id)
+                )
+
+            items[item_id] = DatasetItem(id=item_id, media=image, annotations=annotations)
+
+        return items
+
+    @staticmethod
+    def _lazy_extract_mask(mask, c):
+        return lambda: mask == c
+
+
+class CommonSemanticSegmentationImporter(Importer):
+    @classmethod
+    def build_cmdline_parser(cls, **kwargs):
+        parser = super().build_cmdline_parser(**kwargs)
+        parser.add_argument("--image-prefix", default="", help="Image prefix (default: '')")
+        parser.add_argument("--mask-prefix", default="", help="Mask prefix (default: '')")
+        return parser
+
+    @classmethod
+    def detect(cls, context: FormatDetectionContext) -> None:
+        context.require_file(f"**/{CommonSemanticSegmentationPath.IMAGES_DIR}/**/*.*")
+        context.require_file(f"**/{CommonSemanticSegmentationPath.MASKS_DIR}/**/*.*")
+        context.require_file(f"**/{DATASET_META_FILE}")
```

Comment:
Similarly to the loading code, this doesn't need to search for images and masks everywhere in the tree. Once you have located the meta file, you can take its parent directory and search there.

Reply:
Fixed